imerge/imerge.1
.TH IMERGE 1L
.SH NAME
imerge \- merge two files interactively
.SH SYNOPSIS
.B imerge file1 file2 [mergedfile]
.SH DESCRIPTION
.I Imerge
will read the two files given as first and second parameter,
which will presumably two similar but different versions of a text.
The program will allow you to create a third, merged version
containing "the best" parts of both files.
.PP
The merged version will be written to
.I mergedfile.
It defaults to "imerge.out".
It will (unconditionally) contain all common text.  Differences are
included as described below.
.PP
With a little help from
.IR diff (1),
it will present you the differences in individual
.I hunks.
In the
.I diff
output, text contained in
.I file1
will be marked with a leading '<', while text from
.I file2
is preceeded by a '>'.
One hunk may typically look like this:
.sp
.nf
.B
	3,5c8,9
	<This is some text
	<in file1 and it got
	<somewhat changed.
	\-\-\-
	>This is some corresponding text
	>in the other file.
.fi
.sp
According to the order of the two file arguments, one can think
of "the file on the left" resp. "right".
In the case of additions and deletions,
only one file will be displayed.
.PP
After a single hunk has been presented,
.I imerge
prompts for the action to be taken.
You can either select the passage from
.I file1
or from
.I file2
by answering with a '<' or '>' respectively.
The third possiblity is to invoke an editor on the hunk and merge the
passage by hand.
.I Imerge
looks in the environment variables
.B VISUAL
and
.B EDITOR
in this order to determine the editor for its use,
and will lastly resort to
.IR vi (1).
.PP
Finally, you can execute any shell command by escaping it with a '!'.
.SH "SEE ALSO"
.I Imerge
is nothing else than the program
.I idiff
("interactive diff") from Kernighan & Pike's book
.I "The Unix Programming Environment."
The main goal of the program is to allow merges of concurrently
developed versions.
Other three-way merges can be achieved by Walter F. Tichy's
.I merge
resp.
.I rcsmerge
and of course with Larry Walls famous
.I patch.
You should both of them check out to see if they are better suited to your
specific purposes.
.SH RESTRICTIONS
The output file name must be different from the input files.


ups/CHANGES
@(#)CHANGES	3.35	$Id: CHANGES,v 1.1.1.1 2001/01/27 23:32:36 neitzel Exp $

3.35
----

3.35 has been used on Solaris 2.x, Solaris 8, FreeBSD 3.2,
FreeBSD 4.0 and RedHat Linux 6.2.

See the file README for build instructions.

The URL for the ups web site is:

           http://www.concerto.demon.co.uk/UPS/

It is maintained by Ian Edwards (ian@concerto.demon.co.uk).
It includes a FAQ, html man pages, site listings where ups
can be found, supported architectures, a history of changes
between versions, and other information.

Changes between 3.35 and 3.34
-----------------------------

o  Changes from Russ Browne <russ@amc.com>
    - Support for SC5 compiler:
        Use the compiler supplied demangling library.
        Recognize the bool type, now supported by this compiler.
        Support for new implementation of virtual base classes used
        by this compiler.

    - Corrected look and feel for X clipboard operations:
        Selected text does not vanish when the mouse button is released.
        Instead, it stays selected until another clipboard selection
        is made or the displayed text is changed by some user action.

        Corrected paste from X clipboard to not allow pasting a new-line
        in a text display that does not accept a new-line from the keyboard.

    - Corrected display of source files with dos style line endings to not
      display the \r at the end of each line.

    - Added a keyboard history feature:
        Most regions where UPS accepts data from the keyboard have
        a history of recent entries.  Recent entries can be accessed
        either through the keyboard by typing Control-P or Control-N,
        or through the mouse by pressing a button that brings up a menu
        of recent entries for that item.

    - Added a text-entry button.
        This was primarily so that when entries are accessed through
        the history menu one can complete the entry with the mouse
        rather than shifting to the keyboard to type return or escape.
        Clicking the middle mouse button on this region does a paste.

    - Allow the user to remove the mousehole from the display through X resources.
        If one uses a large font, the typing line area can get
        rather small.  By setting X resources one can retrieve the
        space the mousehole takes up for the typing line instead.

    - Fixed med3 keyboard routines to recognize the most useful control characters.

    - display of 'long double' on SunOS 4.x
    - implement raise/lower on "exec to here".
    - fix to the symbol table problem reported by James Armstrong.
    - stop 'Add Expresion' crash if no debugging info in the file
    - fixes for lower/iconify on run

  Notes from Russ Browne <russ@amc.com> :

   Added X resource and command line option to split the output window
   off as an independent top level X window.

   Arrange that ups can raise itself to the foreground when the debugged
   process breaks or dies.  This behavior can be turned on by a
   menu option under "Windows" (below) or by an X resource or a command
   line option.

   Renamed the old "Snapshot" button to "Windows" and added options to
   raise sister ups windows to the foreground and to make ups raise itself
   to the foreground when the debugged process breaks or dies.

  Some more technical changes:

    Added handling of N_ILDPAD stabs information in elf symbol tables.

    Further tightening up of compiler identification while handling
    the N_OPT stab.

    Treatment of SC5 virtual base classes is working to my satisfaction.
    I removed from the CHANGES file the sentence about this requiring
    a lot more work.

    In split window mode, the typing line '/' and '?' shortcuts always
    search something on the screen where the command is typed.

    Changes in symbol parsing to ensure the the va_type field in
    a var_t object is never NULL.  I reported earlier that I changed
    TypeID to never return NULL; that proved to be a mistake.  Instead,
    I always check that the "rtype" being assigned to v->va_type is
    not NULL.

    Removed the broken mousehole from the source window when it is
    split off a top level window.

    Renamed the X resource to remove the mousehole completely to
    "WantMousehole=no".

    Added X resource and command line option to turn message logging
    on by default.

    Added wn_setdefault() and used it to set several X resources
    from the command line.  X resources so treated appear in the
    command line as [-[no]option], so that an option turned on by an
    X resource can be turned off by the command line, or vice-versa.
    These options are:
       -[no]split[:num] -[no]splitoutput[:num]
       -[no]mousehole -[no]outwin -[no]logging -[no]raise_on_break

    Added support for menu toggle items and used them to combine the
    message logging on/off options into a single option and to add
    the new raise-on-break option.

    When the window containing the display area is iconified all UPS
    windows are iconified.  When that window is de-iconified all windows
    iconfied with it are restored.

    Gave UPS an icon.  As I said earlier, I not a great artist.
    If anyone wants to improve on my work, or replace the icon
    completely, be my guest.  The bitmap for the icon is in ups.c

    Added X resources for the source and output window geometries.
    These are Src.Geometry and Output.Geometry.  Originally, these
    were to be SrcGeometry and OutputGeometry, but the '.' fell out
    of the existing get_xdef routine in wn_init.c and appears to be
    consistent with X conventions for specifying a resource specific
    to an individual window, button, or the like.  In the manual page
    I spelled this out as "Ups.Src.Geometry" because I felt readers
    might not understand that the "Ups." still goes in front of the
    resource name.

o  Changes from Rod Armstrong <rod@san-jose.tt.slb.com>
    - splitting the ups window into two.

imerge/imerge.c
/* idiff: interactive diff */

#include <stdio.h>
#include <ctype.h>

#include <unistd.h>

char *progname;
#define HUGE	10000		/* large number of lines */

main (argc, argv)
int argc;
char *argv[];
{
	FILE *fin, *fout, *f1, *f2, *efopen();
	char buf[BUFSIZ], *mktemp();
	char *outfile;

	progname = argv[0];
	if (argc != 3  && argc != 4) {
		fprintf (stderr, "Usage: %s file1 file2 [mergedfile]\n",
			 progname);
		exit (1);
	}
	f1 = efopen (argv[1], "r");
	f2 = efopen (argv[2], "r");
	outfile = argc==4 ? argv[3] : "imerge.out";
	fout = efopen (outfile, "w");
	sprintf (buf, "diff %s %s", argv[1], argv[2]);
	fin = popen (buf, "r");
	idiff (f1, f2, fin, fout);
	pclose (fin);
	fclose (f1);
	fclose (f2);
	fclose (fout);
	printf ("%s output in file %s\n", progname, outfile);
	exit (0);
}

idiff (f1, f2, fin, fout)	/* process diffs */
FILE *f1, *f2, *fin, *fout;
{
	char *getenv();
	char *tempfile;
	char buf[BUFSIZ], buf2[BUFSIZ];
	FILE *ft, *efopen();
	int cmd, n, from1, to1, from2, to2, nf1, nf2;
	char *editor;

	(editor = getenv("VISUAL")) ||
		(editor = getenv("EDITOR")) ||
			(editor = "vi") ;

	if ( ! (tempfile = tmpnam(NULL))) {
		fprintf (stderr, "%s: tmpnam(3) failed\n", progname);
		exit (1);
	}
	nf1 = nf2 = 0;
	while (fgets (buf, sizeof buf, fin) != NULL) {
		parse (buf, &from1, &to1, &cmd, &from2, &to2);
		n = to1-from1 + to2-from2 + 1; /* #lines from diff */
		if (cmd == 'c')
			n += 2;
		else if (cmd == 'a')
			from1++;
		else if (cmd == 'd')
			from2++;
		printf ("%s", buf);
		while (n-- > 0) {
			fgets (buf, sizeof buf, fin);
			printf ("%s", buf);
		}
		do {
			printf ("? ");
			fflush (stdout);
			fgets (buf, sizeof buf, stdin);
			switch (buf[0]) {
			case '>':
				nskip (f1, to1-nf1);
				ncopy (f2, to2-nf2, fout);
				break;
			case '<':
				nskip (f2, to2-nf2);
				ncopy (f1, to1-nf1, fout);
				break;
			case 'e':
				ncopy (f1, from1-1-nf1, fout);
				nskip (f2, from2-1-nf2);
				ft = efopen (tempfile, "w");
				ncopy (f1, to1+1-from1, ft);
				fprintf (ft, "---\n");
				ncopy (f2, to2+1-from2, ft);
				fclose (ft);
				sprintf (buf2, "%s %s", editor, tempfile);
				system (buf2);
				ft = efopen (tempfile, "r");
				ncopy (ft, HUGE, fout);
				fclose (ft);
				break;
			case '!':
				system (buf+1);
				printf ("!\n");
				break;
			default:
				printf ("%s\n%s\n%s editor '%s'\n%s\n",
					"<	take old hunk",
					">	take new hunk",
					"e	merge hunks in with", editor,
					"!cmd	execute shell cmd");
				break;
			}
		} while (buf[0] != '<' &&
			 buf[0] != '>' &&
			 buf[0] != 'e');
		nf1 = to1;
		nf2 = to2;
	}
	ncopy (f1, HUGE, fout);
	unlink (tempfile);
}

parse (s, pfrom1, pto1, pcmd, pfrom2, pto2)
char *s;
int *pcmd, *pfrom1, *pto1, *pfrom2, *pto2;
{
#define a2i(p) while (isdigit(*s)) p = 10*(p) + *s++ - '0'

	*pfrom1 = *pto1 = *pfrom2 = *pto2 = 0;
	a2i (*pfrom1);
	if (*s == ',') {
		s++;
		a2i (*pto1);
	} else
		*pto1 = *pfrom1;
	*pcmd = *s++;
	a2i (*pfrom2);
	if (*s == ',') {
		s++;
		a2i (*pto2);
	} else
		*pto2 = *pfrom2;
}

nskip (fin, n)
FILE *fin;
int n;
{
	char buf[BUFSIZ];

	while (n-- >0)
		fgets (buf, sizeof buf, fin);
}

ncopy (fin, n, fout)
FILE *fin, *fout;
{
	char buf[BUFSIZ];

	while (n-- > 0) {
		if (fgets (buf, sizeof buf, fin) == NULL)
			return;
		fputs (buf, fout);
	}
}

FILE *efopen (name, mode)
char *name, *mode;
{
	FILE *f;

	if ( ! (f=fopen(name, mode))) {
		fprintf (stderr, "%s: could not open file %s\n",
			 progname, name);
		exit (1);
	} else
		return f;
}
